app/controllers/account_controller.rb

```diff
@@ -153,17 +161,70 @@ def allow_lost_password_recovery?
 
   # Token based account activation
   def activate
-    return redirect_to(home_url) unless Setting.self_registration? && params[:token]
-    token = Token.find_by(action: 'register', value: params[:token].to_s)
-    redirect_to(home_url) && return unless token and !token.expired?
+    token = Token.find_by value: params[:token].to_s
+
+    if token && token.action == 'register' && Setting.self_registration?
+      activate_self_registered token
+    else
+      activate_invited token
+    end
+  end
+
+  def activate_self_registered(token)
     user = token.user
-    redirect_to(home_url) && return unless user.registered?
-    user.activate
-    if user.save
-      token.destroy
-      flash[:notice] = l(:notice_account_activated)
+
+    if not user.registered?
+      if user.active?
+        flash[:notice] = I18n.t(:notice_account_already_activated)
+      else
+        flash[:error] = I18n.t(:notice_activation_failed)
+      end
+
+      redirect_to home_url
+    else
+      user.activate
+
+      if user.save
+        token.destroy
+        flash[:notice] = I18n.t(:notice_account_activated)
+      else
+        flash[:error] = I18n.t(:notice_activation_failed)
+      end
+
+      redirect_to signin_path
+    end
+  end
+
+  def activate_invited(token)
```

Comment:
This is some heavily branched method. Can you take these apart so that we have no more than two nesting levels? Hard for me to follow the flow here.